File: backend/src/controllers/authController.js
const jwt = require('jsonwebtoken');
const User = require('../models/User');

const JWT_SECRET = process.env.JWT_SECRET || 'your-secret-key-change-in-production';
const JWT_EXPIRES_IN = process.env.JWT_EXPIRES_IN || '24h';

// 登录
exports.login = async (req, res) => {
  try {
    const { username, password } = req.body;

    if (!username || !password) {
      return res.status(400).json({
        success: false,
        error: 'Username and password are required',
      });
    }

    // 验证用户
    const user = await User.verifyPassword(username, password);

    if (!user) {
      return res.status(401).json({
        success: false,
        error: 'Invalid username or password',
      });
    }

    // 生成 JWT
    const token = jwt.sign(
      {
        id: user.id,
        username: user.username,
        role: user.role,
      },
      JWT_SECRET,
      { expiresIn: JWT_EXPIRES_IN }
    );

    res.json({
      success: true,
      data: {
        user: {
          id: user.id,
          username: user.username,
          email: user.email,
          role: user.role,
        },
        token,
      },
    });
  } catch (error) {
    console.error('Login error:', error);
    res.status(500).json({
      success: false,
      error: 'Login failed',
    });
  }
};

// 获取当前用户信息
exports.me = async (req, res) => {
  try {
    const user = await User.findById(req.user.id);

    if (!user) {
      return res.status(404).json({
        success: false,
        error: 'User not found',
      });
    }

    res.json({
      success: true,
      data: user,
    });
  } catch (error) {
    console.error('Get user error:', error);
    res.status(500).json({
      success: false,
      error: 'Failed to get user info',
    });
  }
};

// 修改密码
exports.changePassword = async (req, res) => {
  try {
    const { oldPassword, newPassword } = req.body;

    if (!oldPassword || !newPassword) {
      return res.status(400).json({
        success: false,
        error: 'Old password and new password are required',
      });
    }

    // 验证旧密码
    const user = await User.findById(req.user.id);
    const verified = await User.verifyPassword(user.username, oldPassword);

    if (!verified) {
      return res.status(401).json({
        success: false,
        error: 'Invalid old password',
      });
    }

    // 更新密码
    await User.updatePassword(req.user.id, newPassword);

    res.json({
      success: true,
      message: 'Password changed successfully',
    });
  } catch (error) {
    console.error('Change password error:', error);
    res.status(500).json({
      success: false,
      error: 'Failed to change password',
    });
  }
};

// 获取所有用户（仅管理员）
exports.getAllUsers = async (req, res) => {
  try {
    const users = await User.findAll();

    res.json({
      success: true,
      data: users,
    });
  } catch (error) {
    console.error('Get users error:', error);
    res.status(500).json({
      success: false,
      error: 'Failed to get users',
    });
  }
};

// 创建用户（仅管理员）
exports.createUser = async (req, res) => {
  try {
    const { username, email, password, role } = req.body;

    if (!username || !email || !password) {
      return res.status(400).json({
        success: false,
        error: 'Username, email, and password are required',
      });
    }

    // 检查用户名是否已存在
    const existingUser = await User.findByUsername(username);
    if (existingUser) {
      return res.status(409).json({
        success: false,
        error: 'Username already exists',
      });
    }

    // 检查邮箱是否已存在
    const existingEmail = await User.findByEmail(email);
    if (existingEmail) {
      return res.status(409).json({
        success: false,
        error: 'Email already exists',
      });
    }

    // 创建用户
    const user = await User.create(username, email, password, role || 'member');

    res.status(201).json({
      success: true,
      data: user,
    });
  } catch (error) {
    console.error('Create user error:', error);
    res.status(500).json({
      success: false,
      error: 'Failed to create user',
    });
  }
};

// 更新用户（仅管理员）
exports.updateUser = async (req, res) => {
  try {
    const { userId } = req.params;
    const { username, email, role } = req.body;

    const user = await User.update(userId, username, email, role);

    if (!user) {
      return res.status(404).json({
        success: false,
        error: 'User not found',
      });
    }

    res.json({
      success: true,
      data: user,
    });
  } catch (error) {
    console.error('Update user error:', error);
    res.status(500).json({
      success: false,
      error: 'Failed to update user',
    });
  }
};

// 删除用户（仅管理员）
exports.deleteUser = async (req, res) => {
  try {
    const { userId } = req.params;

    // 不能删除自己
    if (parseInt(userId) === req.user.id) {
      return res.status(400).json({
        success: false,
        error: 'Cannot delete yourself',
      });
    }

    const user = await User.delete(userId);

    if (!user) {
      return res.status(404).json({
        success: false,
        error: 'User not found',
      });
    }

    res.json({
      success: true,
      message: 'User deleted successfully',
    });
  } catch (error) {
    console.error('Delete user error:', error);
    res.status(500).json({
      success: false,
      error: 'Failed to delete user',
    });
  }
};

File: backend-csharp/Controllers/AuthController.cs
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using System.Security.Claims;
using iDataflow.Backend.Services;
using iDataflow.Backend.DTOs;

namespace iDataflow.Backend.Controllers
{
    [ApiController]
    [Route("api/[controller]")]
    public class AuthController : ControllerBase
    {
        private readonly UserService _userService;
        private readonly JwtService _jwtService;
        private readonly ILogger<AuthController> _logger;

        public AuthController(UserService userService, JwtService jwtService, ILogger<AuthController> logger)
        {
            _userService = userService;
            _jwtService = jwtService;
            _logger = logger;
        }

        [HttpPost("login")]
        public async Task<IActionResult> Login([FromBody] LoginRequest request)
        {
            try
            {
                if (string.IsNullOrEmpty(request.Username) || string.IsNullOrEmpty(request.Password))
                {
                    return BadRequest(new ApiResponse
                    {
                        Success = false,
                        Error = "Username and password are required"
                    });
                }

                var user = await _userService.VerifyPasswordAsync(request.Username, request.Password);

                if (user == null)
                {
                    return Unauthorized(new ApiResponse
                    {
                        Success = false,
                        Error = "Invalid username or password"
                    });
                }

                var token = _jwtService.GenerateToken(user);

                return Ok(new ApiResponse
                {
                    Success = true,
                    Data = new LoginResponse
                    {
                        User = new UserDto
                        {
                            Id = user.Id,
                            Username = user.Username,
                            Email = user.Email,
                            Role = user.Role
                        },
                        Token = token
                    }
                });
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Login error for user {Username}", request.Username);
                return StatusCode(500, new ApiResponse
                {
                    Success = false,
                    Error = "Login failed"
                });
            }
        }

        [HttpGet("me")]
        [Authorize]
        public async Task<IActionResult> Me()
        {
            try
            {
                var userId = int.Parse(User.FindFirst("id")?.Value ?? "0");
                var user = await _userService.FindByIdAsync(userId);

                if (user == null)
                {
                    return NotFound(new ApiResponse
                    {
                        Success = false,
                        Error = "User not found"
                    });
                }

                return Ok(new ApiResponse
                {
                    Success = true,
                    Data = new UserDto
                    {
                        Id = user.Id,
                        Username = user.Username,
                        Email = user.Email,
                        Role = user.Role
                    }
                });
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Get user error");
                return StatusCode(500, new ApiResponse
                {
                    Success = false,
                    Error = "Failed to get user info"
                });
            }
        }

        [HttpPost("change-password")]
        [Authorize]
        public async Task<IActionResult> ChangePassword([FromBody] ChangePasswordRequest request)
        {
            try
            {
                if (string.IsNullOrEmpty(request.OldPassword) || string.IsNullOrEmpty(request.NewPassword))
                {
                    return BadRequest(new ApiResponse
                    {
                        Success = false,
                        Error = "Old password and new password are required"
                    });
                }

                var userId = int.Parse(User.FindFirst("id")?.Value ?? "0");
                var username = User.FindFirst("username")?.Value ?? "";

                // Verify old password
                var user = await _userService.VerifyPasswordAsync(username, request.OldPassword);
                if (user == null)
                {
                    return Unauthorized(new ApiResponse
                    {
                        Success = false,
                        Error = "Invalid old password"
                    });
                }

                // Update password
                await _userService.UpdatePasswordAsync(userId, request.NewPassword);

                return Ok(new ApiResponse
                {
                    Success = true,
                    Data = new { Message = "Password changed successfully" }
                });
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Change password error");
                return StatusCode(500, new ApiResponse
                {
                    Success = false,
                    Error = "Failed to change password"
                });
            }
        }

        [HttpGet("users")]
        [Authorize(Roles = "admin")]
        public async Task<IActionResult> GetAllUsers()
        {
            try
            {
                var users = await _userService.GetAllUsersAsync();

                var userDtos = users.Select(u => new UserDto
                {
                    Id = u.Id,
                    Username = u.Username,
                    Email = u.Email,
                    Role = u.Role
                }).ToList();

                return Ok(new ApiResponse
                {
                    Success = true,
                    Data = userDtos
                });
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Get users error");
                return StatusCode(500, new ApiResponse
                {
                    Success = false,
                    Error = "Failed to get users"
                });
            }
        }

        [HttpPost("users")]
        [Authorize(Roles = "admin")]
        public async Task<IActionResult> CreateUser([FromBody] CreateUserRequest request)
        {
            try
            {
                if (string.IsNullOrEmpty(request.Username) || string.IsNullOrEmpty(request.Email) || string.IsNullOrEmpty(request.Password))
                {
                    return BadRequest(new ApiResponse
                    {
                        Success = false,
                        Error = "Username, email, and password are required"
                    });
                }

                // Check if user exists
                if (await _userService.UserExistsAsync(request.Username, request.Email))
                {
                    return Conflict(new ApiResponse
                    {
                        Success = false,
                        Error = "Username or email already exists"
                    });
                }

                var user = await _userService.CreateUserAsync(request.Username, request.Email, request.Password, request.Role ?? "member");

                return CreatedAtAction(nameof(GetAllUsers), new ApiResponse
                {
                    Success = true,
                    Data = new UserDto
                    {
                        Id = user.Id,
                        Username = user.Username,
                        Email = user.Email,
                        Role = user.Role
                    }
                });
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Create user error");
                return StatusCode(500, new ApiResponse
                {
                    Success = false,
                    Error = "Failed to create user"
                });
            }
        }

        [HttpPut("users/{userId}")]
        [Authorize(Roles = "admin")]
        public async Task<IActionResult> UpdateUser(int userId, [FromBody] UpdateUserRequest request)
        {
            try
            {
                var user = await _userService.UpdateUserAsync(userId, request.Username, request.Email, request.Role);

                if (user == null)
                {
                    return NotFound(new ApiResponse
                    {
                        Success = false,
                        Error = "User not found"
                    });
                }

                return Ok(new ApiResponse
                {
                    Success = true,
                    Data = new UserDto
                    {
                        Id = user.Id,
                        Username = user.Username,
                        Email = user.Email,
                        Role = user.Role
                    }
                });
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Update user error");
                return StatusCode(500, new ApiResponse
                {
                    Success = false,
                    Error = "Failed to update user"
                });
            }
        }

        [HttpDelete("users/{userId}")]
        [Authorize(Roles = "admin")]
        public async Task<IActionResult> DeleteUser(int userId)
        {
            try
            {
                var currentUserId = int.Parse(User.FindFirst("id")?.Value ?? "0");

                if (userId == currentUserId)
                {
                    return BadRequest(new ApiResponse
                    {
                        Success = false,
                        Error = "Cannot delete yourself"
                    });
                }

                var deleted = await _userService.DeleteUserAsync(userId);

                if (!deleted)
                {
                    return NotFound(new ApiResponse
                    {
                        Success = false,
                        Error = "User not found"
                    });
                }

                return Ok(new ApiResponse
                {
                    Success = true,
                    Data = new { Message = "User deleted successfully" }
                });
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Delete user error");
                return StatusCode(500, new ApiResponse
                {
                    Success = false,
                    Error = "Failed to delete user"
                });
            }
        }
    }
}